File: bridge/DelayNode.cpp
#include "./dart_api/dart_api.h"
#include "LabSound/LabSound.h"
#include "KeepNode.cpp"
using namespace lab;

DART_EXPORT int createDelayNode(AudioContext* context) {
    auto node = std::make_shared<DelayNode>(*context);
    return keepNode(node);
}

DART_EXPORT int DelayNode_delayTime(int nodeId) {
    auto node = std::static_pointer_cast<DelayNode>(getNode(nodeId));
    return node ? keepAudioSetting(nodeId, 0, node->delayTime()) : -1;
}

File: lab_sound_ffi/lib/src/lab_sound/delay_node.dart
import 'audio_context.dart';
import 'audio_node.dart';
import 'audio_setting.dart';
import 'lab_sound.dart';

class DelayNode extends AudioNode {
  late AudioSettingFloat delayTime;
  DelayNode(AudioContext ctx, {double maxDelayTime = 2.0})
      : super(ctx, LabSound().createDelayNode(ctx.pointer, maxDelayTime)) {
    delayTime = AudioSettingFloat(
        this.ctx, nodeId, LabSound().DelayNode_delayTime(nodeId));
  }

  DelayNode.fromId(AudioContext ctx, int id) : super(ctx, id) {
    delayTime = AudioSettingFloat(
        this.ctx, nodeId, LabSound().DelayNode_delayTime(nodeId));
  }
}
